scoped-mutex/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+#![no_std]
+
+use core::cell::UnsafeCell;
+use core::sync::atomic::{AtomicBool, Ordering};
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`super::Mutex`] instead,
+/// which is generic over a RawMutex and contains the protected data.
+///
+/// Note that, unlike other mutexes, implementations only guarantee no
+/// concurrent access from other threads: concurrent access from the current
+/// thread is allowed. For example, it's possible to lock the same mutex multiple times reentrantly.
+///
+/// Therefore, locking a `RawMutex` is only enough to guarantee safe shared (`&`) access
+/// to the data, it is not enough to guarantee exclusive (`&mut`) access.
+///
+/// # Safety
+///
+/// RawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+pub unsafe trait RawMutex {
+    /// Create a new `RawMutex` instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
+
+    /// Lock this `RawMutex`.
+    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
```

Comment:
Oh, one other thing: I'd really like a 
```suggestion
    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
    
    /// Try to lock this `RawMutex`, without blocking.
    ///
    /// If the lock can be acquired without blocking, this method calls `f` and returns the
    /// result; otherwise, it returns `None`.
    fn try_lock<R>(&self, f: impl FnOnce() -> R) -> Option<R>;
```

I generally use `try_lock` methods in the `fmt::Debug` implementations for my synchronization primitives in `maitake-sync`, and it would be nice to be able to implement `Debug` with scoped raw mutex impls...

Reply:
One other note: what do you think about calling this method either `with_lock` or `locked`, instead of `lock`? Again, I'm mainly thinking about the case of types that provide both scoped _and_ RAII APIs: in that case, the methods on the Mutex type will need to have different names, and I'd like to be able to reserve `Mutex::lock` for a RAII API for consistency with `std`. It would be nice for the mutex type methods to have the same name as the raw mutex API.

In my current `maitake-sync` branch, I called this `with_lock`, but thinking about it some more, I'm starting to wonder if I prefer `locked`. The rationale is that I also have _constructors_ that include `with_` in their names (`Mutex::with_raw_mutex` etc) and it seems slightly unfortunate to use the same prefix in a constructor and a method...though on the other hand, I could change the constructors to `new_with_...`. What do you think?